### ISSUE TITLE
Recent Kokkos changes require -std=c++11 flag with CUDA 8.0 & GCC 5.3

### ISSUE THREAD

**Issue by mhoemmen** [trilinos/Trilinos#1630]
@trilinos/framework 

It looks like C++11 is not being enabled by default in a CUDA 8.0 + GCC 5.3 build with TpetraCore enabled.  When I omit `-std=c++11` from `CMAKE_CXX_FLAGS` in a CUDA 8.0 + GCC 5.3 build with TpetraCore enabled, I get the following configure-time error:
```
-- MPI_EXEC='/projects/sems/install/rhel6-x86_64/kokkos/compiler/gcc/5.3.0/openmpi/2.0.1/cuda/8.0.44/bin/mpiexec'
-- The C compiler identification is GNU 5.3.0
-- Check for working C compiler: /projects/sems/install/rhel6-x86_64/kokkos/compiler/gcc/5.3.0/openmpi/2.0.1/cuda/8.0.44/bin/mpicc
-- Check for working C compiler: /projects/sems/install/rhel6-x86_64/kokkos/compiler/gcc/5.3.0/openmpi/2.0.1/cuda/8.0.44/bin/mpicc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- CMAKE_C_COMPILER_ID='GNU'
-- CMAKE_C_COMPILER_VERSION='5.3.0'
-- The CXX compiler identification is unknown
-- Check for working CXX compiler: /projects/sems/install/rhel6-x86_64/kokkos/compiler/gcc/5.3.0/openmpi/2.0.1/cuda/8.0.44/bin/mpicxx
-- Check for working CXX compiler: /projects/sems/install/rhel6-x86_64/kokkos/compiler/gcc/5.3.0/openmpi/2.0.1/cuda/8.0.44/bin/mpicxx -- broken
CMake Error at /projects/sems/install/rhel6-x86_64/sems/utility/cmake/3.3.2/share/cmake-3.3/Modules/CMakeTestCXXCompiler.cmake:54 (message):
  The C++ compiler
  "/projects/sems/install/rhel6-x86_64/kokkos/compiler/gcc/5.3.0/openmpi/2.0.1/cuda/8.0.44/bin/mpicxx"
  is not able to compile a simple test program.

  It fails with the following output:

   Change Dir: .../Trilinos/CHECKIN-CUDA-8.0/MPI_DEBUG_REAL/CMakeFiles/CMakeTmp

  Run Build Command:"/usr/bin/gmake" "cmTC_64169/fast"

  /usr/bin/gmake -f CMakeFiles/cmTC_64169.dir/build.make
  CMakeFiles/cmTC_64169.dir/build

  gmake[1]: Entering directory
  `.../Trilinos/CHECKIN-CUDA-8.0/MPI_DEBUG_REAL/CMakeFiles/CMakeTmp'

  Building CXX object CMakeFiles/cmTC_64169.dir/testCXXCompiler.cxx.o
/projects/sems/install/rhel6-x86_64/kokkos/compiler/gcc/5.3.0/openmpi/2.0.1/cuda/8.0.44/bin/mpicxx
  -Wall -expt-extended-lambda -o
  CMakeFiles/cmTC_64169.dir/testCXXCompiler.cxx.o -c
  .../Trilinos/CHECKIN-CUDA-8.0/MPI_DEBUG_REAL/CMakeFiles/CMakeTmp/testCXXCompiler.cxx

  /projects/sems/install/rhel6-x86_64/kokkos/compiler/cuda/8.0.44/bin/..//include/cuda_runtime.h:140:1:
  warning: identifier ‘static_assert’ is a keyword in C++11
  [-Wc++0x-compat]

  /projects/sems/install/rhel6-x86_64/kokkos/compiler/cuda/8.0.44/bin/..//include/cuda_runtime.h:156:1:
  warning: identifier ‘decltype’ is a keyword in C++11 [-Wc++0x-compat]

    *
   ^

  /projects/sems/install/rhel6-x86_64/kokkos/compiler/cuda/8.0.44/bin/..//include/cuda_runtime.h:133:29:
  error: template argument 1 is invalid

   #if defined(__cplusplus) && !defined(__CUDACC_RTC__)
                               ^

  /projects/sems/install/rhel6-x86_64/kokkos/compiler/cuda/8.0.44/bin/..//include/cuda_runtime.h:138:32:
  warning: variadic templates only available with -std=c++11 or -std=gnu++11

   *                                                                              *
                                  ^

  /projects/sems/install/rhel6-x86_64/kokkos/compiler/cuda/8.0.44/bin/..//include/cuda_runtime.h:140:15:
  error: expected identifier before ‘sizeof’
/projects/sems/install/rhel6-x86_64/kokkos/compiler/cuda/8.0.44/bin/..//include/cuda_runtime.h:140:15:
  error: expected ‘,’ or ‘...’ before ‘sizeof’

  /projects/sems/install/rhel6-x86_64/kokkos/compiler/cuda/8.0.44/bin/..//include/cuda_runtime.h:140:104:
  error: ISO C++ forbids declaration of ‘static_assert’ with no type
  [-fpermissive]

  /projects/sems/install/rhel6-x86_64/kokkos/compiler/cuda/8.0.44/bin/..//include/cuda_runtime.h:145:19:
  warning: variadic templates only available with -std=c++11 or -std=gnu++11

  /projects/sems/install/rhel6-x86_64/kokkos/compiler/cuda/8.0.44/bin/..//include/cuda_runtime.h:146:18:
  warning: variadic templates only available with -std=c++11 or -std=gnu++11

  /projects/sems/install/rhel6-x86_64/kokkos/compiler/cuda/8.0.44/bin/..//include/cuda_runtime.h:148:49:
  warning: variadic templates only available with -std=c++11 or -std=gnu++11
....
```
When I add `-std=c++11` to `CMAKE_CXX_FLAGS`, I am able to configure Trilinos and build TpetraCore library, tests, and examples.

**Comment by mhoemmen** [comment]
Here are my relevant CMake options:
```
-D Trilinos_ENABLE_EXPLICIT_INSTANTIATION:BOOL=ON
-D BUILD_SHARED_LIBS:BOOL=ON
-D Trilinos_ENABLE_OpenMP:BOOL=ON
-D Trilinos_SHOW_DEPRECATED_WARNINGS:BOOL=ON
-D Trilinos_ENABLE_Fortran:BOOL=OFF
-D CMAKE_Fortran_LINK_EXECUTABLE:FILEPATH=""
-D TPL_ENABLE_CUDA:BOOL=ON
  -D Kokkos_ENABLE_Cuda_UVM:BOOL=ON
  -D Tpetra_INST_CUDA:BOOL=ON
  -D Kokkos_ENABLE_Cuda_Lambda:BOOL=ON
  -D Kokkos_ENABLE_CXX11_DISPATCH_LAMBDA:BOOL=ON
-D CMAKE_CXX_FLAGS:STRING="-Wall -expt-extended-lambda"
-D TPL_ENABLE_MPI:BOOL=ON
-D CMAKE_BUILD_TYPE:STRING=DEBUG
-D Kokkos_ENABLE_DEBUG:BOOL=ON
-D Teuchos_ENABLE_DEBUG:BOOL=ON
-D CMAKE_CXX_FLAGS_DEBUG:STRING="-Os"
-D Tpetra_INST_SERIAL:BOOL=OFF
-D Teuchos_ENABLE_COMPLEX:BOOL=OFF
-D Trilinos_ENABLE_COMPLEX_FLOAT:BOOL=OFF
-D Trilinos_ENABLE_COMPLEX_DOUBLE:BOOL=OFF
```
I am building using the SEMS modules.  Here are the modules I have loaded:
```
$ module list
Currently Loaded Modulefiles:
  1) sems-env                    3) sems-cmake/3.3.2            5) kokkos-cuda/8.0.44
  2) kokkos-env                  4) sems-gcc/5.3.0              6) kokkos-openmpi/2.0.1/cuda
```
Here is the output of `$ mpicxx --version`:
```
$ mpicxx --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2016 NVIDIA Corporation
Built on Sun_Sep__4_22:14:01_CDT_2016
Cuda compilation tools, release 8.0, V8.0.44
```

**Comment by mhoemmen** [comment]
This may be relevant to the recent Kokkos snapshot.
@crtrott @dsunder @ibaned @tjfulle @brian-kelley @trilinos/tpetra

**Comment by ibaned** [comment]
This is more of a TriBITS or configuration issue? No one should be trying to build Kokkos without C++11, and the Kokkos CMake code is not the one responsible for enabling and adding the C++11 flag

**Comment by tjfulle** [comment]
I just ran in to this problem today.  Like @mhoemmen, adding `-std=c++11` to `CMAKE_CXX_FLAGS` gets by the error.  The combination of

```
-D Trilinos_ENABLE_CXX11=ON
-D Trilinos_CXX11_FLAGS:STRING=-std=c++11
```

does not seem to put `-std=c++11` in the c++ command.

I don't know when things changed, but I was able to configure/compile about a month ago without adding `-std=c++11` to the `CMAKE_CXX_FLAGS`.

Perhaps unrelated, with the current setup (adding `-std=c++11` to `CMAKE_CXX_FLAGS`) I get hundreds (thousands?) of these little guys:

```
cc1: warning: command line option ‘-std=c++11’ is valid for C++/ObjC++ but not for C
```

**Comment by bartlettroscoe** [comment]
I would need to see the full CMakeCache.txt file, the full cmake STDOUT and the full cmake commandline.  That should be enough.  Otherwise, I would need to be on a machine where I can reproduce the problem myself to help.

**Comment by jhux2** [comment]
> Perhaps unrelated, with the current setup (adding -std=c++11 to CMAKE_CXX_FLAGS) I get hundreds (thousands?) of these little guys:
> 
> cc1: warning: command line option ‘-std=c++11’ is valid for C++/ObjC++ but not for C

@tjfulle I believe that is resulting from the C++ option being passed on to nvcc, which doesn't understand it.  I'd be really interested in knowing a general method of preventing C++ compile options through.  You can see other instances on the dashboard, e.g., [here](https://testing.sandia.gov/cdash/viewBuildError.php?type=1&buildid=3083359).

**Comment by crtrott** [comment]
The problem is that you add --expt-extended-lambda and then CMake tries to run a test of the compiler using that flag without adding the C++11 flag. That is a CMake issue ...

**Comment by bartlettroscoe** [comment]
This will hopefully be solved when #1400 is complete (no ETA on that at this point).

**Comment by bartlettroscoe** [comment]
Wow, C++14 is not required and working for Trilinos.  Closing as complete.
